### ISSUE TITLE
The CMake does not support multi-file kernel builds. 

### ISSUE THREAD

**Issue by lfield** [illuhad/hipSYCL#277]
The CMake CUDA build `-DHIPSYCL_PLATFORM=cuda` does not support building kernels using multiple translation units. For example, if I have a main.cpp with extras.cpp and extras.h containing a complex number class I get the following error:
`ptxas fatal   : Unresolved extern function '_ZNK6extras7complex4realEv'`
The build works directly from the command line:
`/opt/hipSYCL/bin/syclcc --hipsycl-platform=cpu -O0 -g  -DHIPSYCL_PLATFORM=cuda -DHIPSYCL_GPU_ARCH=sm_70 -DHIPSYCL_CUDA_PATH=/opt/hipSYCL/cuda main.cpp extras.cpp`

**Comment by illuhad** [comment]
This has nothing to do with cmake.

If you look closely at your command line, you see that you are actually compiling for CPU since you pass `--hipsycl-platform=cpu`. The `-DHIPSYCL_XYZ=...` is only for cmake. Because `syclcc` behaves like a regular C++ compiler you are actually only setting preprocessor definitions...

If you compile for CPU, you can do anything that's possible in C++. If you compile for GPU however, restrictions apply for the code that's allowed inside kernels as outlined in the SYCL specification (note that restrictions in kernel code is something all heterogeneous programming models have in common, SYCL, CUDA, OpenCL etc).

If you compile for GPU, the SYCL specification requires that the definition of all functions used inside a kernel (i.e. the entire call tree) be visible at the site of kernel invocation. This means that dividing code in header and cpp file is *not* allowed for kernel code.
The reason for this is that in SYCL you do not mark functions explicitly to be compiled for device. Therefore, the only way the SYCL compiler has to determine whether a function has to be compiled for device is by going through all function calls inside a kernel to build the complete call tree, and check which functions are are required by a kernel. If parts of that call tree are defined in an external cpp file, the SYCL compiler cannot build see the full call tree, so parts of required code will not be compiled for device because the SYCL compiler cannot know that they are needed.
This then of course leads to linker errors as you have seen.
The SYCL specification mentions as *optional* feature the solution that in such cases a programmer can mark functions at the interface of compilation units explicitly with the `SYCL_EXTERNAL` attribute. However, this is an optional in the SYCL specification and we cannot support that in hipSYCL because of limitations in HIP/CUDA clang.

In summary, not a bug and hipSYCL behaves as defined by the specification :)

**Comment by lfield** [comment]
Thanks. I was not aware of this limitation. Would it be possible for cmake to merge all the sources into one file before compiling?

**Comment by illuhad** [comment]
I'm not a cmake expert, but I have never heard of any such feature.. I think you can probably execute arbitrary commands from cmake, so you could hack this into your cmake file I suppose. In any case I don't think this is such a good idea. Merging cpp files into one in an automated process is probably going to break things at some point. After all, you are merging code that was never intended to be compiled as one compilation unit, so you might run into name collisions etc. This is clearly not how to C++...
You can ask in some cmake support forum, but for these reasons I'm skeptical that the cmake people would see much value in this feature.
